src/main/scala/calc/Compiler.scala

```diff
@@ -62,21 +61,212 @@ object Compiler {
   }
 
   /** Compile an expression tree into an IR `Tree`, which is an expression
-   *  that evaluates to the result of the tree.
-   *
-   *  This is the main method you have to implement.
-   */
-  def compileExpr(tree: Tree): irt.Tree = {
-    // TODO
+    *  that evaluates to the result of the tree.
+    *
+    *  This is the main method you have to implement.
+    */
+  def compileExpr(tree: Tree, etaExpansion: Boolean = true): irt.Tree = {
     implicit val pos = tree.pos
 
-    tree match {
-      case Literal(value) =>
-        irt.DoubleLiteral(value)
+    val foreignEnv = Foreign.staticJavaMethods(
+      clsName = "java.lang.Math",
+      importedMethods = List(
+        ("sin", TFun(1)),
+        ("cos", TFun(1)),
+        ("sqrt", TFun(1)),
+        ("pow", TFun(2))
+      )
+    )
+
+    // Phase 1: Convert AST to Typed AST
+    implicit val typeEnv = Typer.emptyEnv ++ foreignEnv
+    val (typedAst, _) = Typer.inferType(tree)
+
+    // Phase 2: Compile Typed AST to IR
+    val ir = expr(typedAst)
+
+    // Phase 3: Prepend Eta expansion of foreign method to IR
+    if (etaExpansion) {
+      Foreign.prependEtaExpansion(ir, foreignEnv)
+    } else {
+      ir
+    }
+  }
+
+  def expr(t: TreeT): irt.Tree = {
+    t match {
+      case t: IdentT => ident(t)
+      case t: LiteralT => literal(t)
+      case t: BinaryOpT => binaryOp(t)
+      case t: LetT => letBinding(t)
+      case t: IfT => ifElse(t)
+      case t: ClosureT => closure(t)
+      case t: CallT => call(t)
+      case t: ForeignCallT => foreignCall(t)
+      case t: SelectT => select(t)
+    }
+  }
+
+  def ident(t: IdentT): irt.Tree = { implicit val pos = t.pos
+    irt.VarRef(irt.Ident(t.name))(t.tpe.irtype)
+  }
+
+  def literal(t: LiteralT) = { implicit val pos = t.pos
+    irt.DoubleLiteral(t.value)
+  }
+
+  def binaryOp(t: BinaryOpT) = { implicit val pos = t.pos
+    val irOp = operatorToIR(t.op) getOrElse (throw UnknownOperator(pos, t.op))
+    irt.BinaryOp(irOp, expr(t.lhs), expr(t.rhs))
+  }
+
+  def letBinding(t: LetT) = { implicit val pos = t.pos
+    val body = expr(t.body)
+    t.value match {
+      case tree:ClosureT =>
+        // Create empty object with name f__ref
+        val emptyObject = irt.JSObjectConstr(List())
+        val bindingName = t.name.name
+        val objectName = irt.Ident(bindingName + "__ref")
+        val objectDef = irt.VarDef(objectName, irtpe.AnyType, false, emptyObject)
```

Comment:
You can find the definitions of the `Ref` classes here: https://github.com/scala-js/scala-js/blob/master/library-aux/src/main/scala/scala/runtime/RefTypes.scala